src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyNode.cs

```diff
@@ -193,47 +167,106 @@ public virtual string Alias
+        public ImmutableHashSet<IDependencyNode> Children
         {
             get
             {
-                return _children;
+                lock (_children)
+                {
+                    return _children;
+                }
             }
         }
 
         public virtual bool HasChildren
         {
             get
             {
-                return Children.Any();
+                return Children.Count > 0;
             }
         }
 
+        /// <summary>
+        /// Adds a child to the node
+        /// </summary>
+        /// <param name="childNode">child node to be added. Should not be null.</param>
         public void AddChild(IDependencyNode childNode)
         {
             Requires.NotNull(childNode, nameof(childNode));
 
-            _children.Add(childNode);
+            lock(_children)
+            {
+                _children = _children.Add(childNode);
+            }
+        }
+
+        /// <summary>
+        /// Adds children to the node
+        /// </summary>
+        /// <param name="children">child nodes to be added. Should not be null.</param>
+        public void AddChildren(IEnumerable<IDependencyNode> children)
+        {
+            Requires.NotNull(children, nameof(children));
+
+            lock (_children)
+            {
+                var builder = _children.ToBuilder();
+                foreach (var child in children)
+                {
+                    builder.Add(child);
+                }
+
+                _children = builder.ToImmutableHashSet();
+            }
         }
 
+        /// <summary>
+        /// Removes child from node's children.
+        /// </summary>
+        /// <param name="childNode">Node to be removed if it belongs to children. Should not be null.</param>
         public void RemoveChild(IDependencyNode childNode)
         {
             Requires.NotNull(childNode, nameof(childNode));
 
-            if (_children.Contains(childNode))
+            lock (_children)
+            {
+                if (_children.Contains(childNode))
+                {
+                    _children = _children.Remove(childNode);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Removes all children from node
+        /// </summary>
+        public void RemoveAllChildren()
+        {
+            lock (_children)
+            {
+                _children = _children.Clear();
+            }
+        }
+
+        /// <summary>
+        /// Changes some properties of the node
+        /// </summary>
+        public void SetProperties(string caption = null)
+        {
+            if (caption != null)
             {
-                _children.Remove(childNode);
+                Caption = caption;
             }
         }
 
         public override int GetHashCode()
         {
-            return unchecked(Id.GetHashCode());
+            return unchecked(Id.GetHashCode() + Resolved.GetHashCode());
```

Comment:
Resolved can not be modified so it would be ok, however your comment is good since i actually don't need resolved here, since in most cases we need to compare node ids. In case when i need Resolved comparison also , i ill create another comparer.

---
In reply to: [99461614](https://github.com/dotnet/roslyn-project-system/pull/1469#discussion_r99461614) [](ancestors = 99461614)